gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/transactions/TransactionDUnit.java

```diff
@@ -145,6 +146,140 @@ public static void resetObservers() {
     }
   }
 
+  public void testStateFlushWithTXHA() throws Exception {
+
+    java.sql.Connection conn = TestUtil.jdbcConn;
+    conn.setTransactionIsolation(getIsolationLevel());
+    Statement st = conn.createStatement();
+    conn.setAutoCommit(false);
+    st.execute("Create table trade.t1 (c1 int not null , c2 int not null, "
+        + "primary key(c1)) partition by primary key redundancy 1 buckets 3 persistent asynchronous " + getSuffix());
+    for(int i=0;i<226;i++) {
+      st.execute("insert into trade.t1 values (" + i + "," + i+")");
+    }
+    conn.commit();
+
+    VM server1 = serverVMs.get(0);
+    VM server3 = serverVMs.get(2);
+
+    server1.invoke(new SerializableRunnable() {
+      @Override
+      public void run() {
+
+        Object waitLock = new Object();
+        final boolean[] executed = new boolean[2];
+        Thread otherTx = new Thread(new Runnable() {
+          @Override
+          public void run() {
+            try {
+              Connection otherConn = TestUtil.getConnection();
+              otherConn.setTransactionIsolation(getIsolationLevel());
+              otherConn.setAutoCommit(false);
+              Statement otherSt = otherConn.createStatement();
+              try {
+                for(int i=226;i<339;i++) {
+                  otherSt.execute("insert into trade.t1 values (" + i + "," + i+")");
+                }
+                while (!executed[0]) {
+                  synchronized (waitLock) {
+                    executed[1] = true;
+                    waitLock.wait();
+                  }
+                }
+                Misc.getI18NLogWriter().info(LocalizedStrings.DEBUG, "Going to inesrt for one bucket");
+
+                PartitionedRegion r = (PartitionedRegion)Misc.getRegionForTable("TRADE.T1", true);
+
+                for(int i:r.getLocalBucketsListTestOnly()) {
+                  Misc.getI18NLogWriter().info(LocalizedStrings.DEBUG, "Going to inesrt for bucket id " + i);
+                  otherSt.execute("insert into trade.t1 values (" + (i + 113 * 4) + ", 20)");
+                }
+
+                otherConn.commit();
+              } catch (Exception e) {
+                otherConn.rollback();
+                e.printStackTrace();
+              }
+            } catch (Exception e) {
+              e.printStackTrace();
+            } finally {
+            }
+          }
+        });
+
+        otherTx.start();
+        while (!executed[1]) {
+          try {
+            Thread.sleep(100);
+          } catch (InterruptedException e) {
+            e.printStackTrace();
+          }
+        }
+        final TransactionObserver observer = new TransactionObserverAdapter() {
+
+          @Override
+          public void beforePerformOp(TXStateProxy tx) {
+            Misc.getI18NLogWriter().info(LocalizedStrings.DEBUG, "before Op : Waiting for 10 sec");
+            try {
+              Thread.sleep(60000);
```

Comment:
Test looks good. But please try and avoid sleeps and use actual latches and booleans so that timing issues don't fail the test unnecessarily.

Reply:
Will change the tests later. Checking it in for regression.